syntax = "proto2";

package communication_commandes;

message Command {
  enum Command_types {
    GOTO = 0;
    PICKUP = 1;
  }
  optional Command_types command = 1;
  
  //parameters below used only if goto command
  optional float dir = 2;
  optional float speed = 3;
  optional float time =4;
  
}

message State {

	//info about robots
	repeated Robot robots = 1;
	
	message Robot {
		optional float x = 1;
		optional float y = 2;
		optional float battery = 3;
		optional bool is_moving = 4;
	
	}
	
	//info about packages
	repeated Package packages = 2;  
	
	message Package {
		optional Location location = 1;
		repeated int32 processes_list = 2;
		optional float delivery_time = 3;
	}
	
	message Location {//contains location of a package
		enum Location_Type {
			CARRIED = 0;
			IN_STAND = 1;
			OUT_STAND = 2;
		}		
		enum Parent_Type {
			ROBOT = 0;
			MACHINE = 1;
			ARRIVAL = 2;
			DELIVERY = 3;
		}
		optional Location_Type location_type = 1;
		optional Parent_Type parent_type = 2;
		optional int32 parent_id = 3;
	}

	//info about machines
	//repeated Machine machines = 3;
	
}



message Environment_Description {
 //to be sent at the start of the simulation to give information about the disposition of the world
	repeated Machine machines = 1;
	
		
	message Machine {
		optional Area_Description input_area = 1;
		optional Area_Description output_area = 2;
		optional int32 input_size = 3;
		optional int32 output_size = 4;
		optional int32 processes_list = 5;
	}
	
	
	optional Area_Description arrival_area = 2;
	
	optional Area_Description delivery_area = 3;
	
	message Area_Description {//used to describe where are physically located the various belts the robot can int32eract with
	optional float x = 1;
	optional float y = 2;
	optional float width = 3;
	optional float height = 4;
}
}



